src/mirage/mirage.mli

```diff
@@ -84,23 +96,99 @@ module Make
     [ standard_status
     | `Status of int ]
 
+  val random : int -> string
+
+  val log : ('a, Format.formatter, unit, unit) format4 -> 'a
+
+  type ('a, 'b) conditional_log =
+    ((?request:request ->
+     ('a, Format.formatter, unit, 'b) format4 -> 'a) -> 'b) ->
+      unit
+
+  type log_level = [ `Error | `Warning | `Info | `Debug ]
+
+  val error     : ('a, unit) conditional_log
+  val warning   : ('a, unit) conditional_log
+  val info      : ('a, unit) conditional_log
+  val debug     : ('a, unit) conditional_log
+
   val html : ?status:status -> ?code:int -> ?headers:(string * string) list -> string -> response Lwt.t
 
   val param : string -> request -> string
 
+  type csrf_result =
+    [ `Ok
+    | `Expired of float
+    | `Wrong_session
+    | `Invalid ]
+
+  val csrf_token : ?valid_for:float -> request -> string
+  val verify_csrf_token : request -> string -> csrf_result Lwt.t
+
+  type 'a form_result =
+    [ `Ok            of 'a
+    | `Expired       of 'a * float
+    | `Wrong_session of 'a
+    | `Invalid_token of 'a
+    | `Missing_token of 'a
+    | `Many_tokens   of 'a
+    | `Wrong_content_type ]
+
+  type multipart_form =
+    (string * ((string option * string) list)) list
+
+  val form : request -> (string * string) list form_result Lwt.t
+  val multipart : request -> multipart_form form_result Lwt.t
+
+  val form_tag :
+    ?method_:method_ ->
+    ?target:string ->
+    ?enctype:[ `Multipart_form_data ] ->
+    ?csrf_token:bool ->
+      action:string -> request -> string
+
+  val lowercase_headers : middleware
+  val content_length : middleware
+
   val logger : middleware
   val router : route list -> middleware
 
   val get : string -> handler -> route
   val not_found : handler
 
+  type error =
+    { condition :
+        [ `Response of response
+        | `String of string
+        | `Exn of exn ]
+    ; layer :
```

Comment:
Not a blocker for merging, but: can you use the `.ocamlformat` from this repo? Or is this formatted manually? The `;`-first style AFAIK dates from when OCaml didn't have trailing separators, and mimics languages like Elm which don't have them. It's very unlike JS/JSON style, is not Web-friendly.  If all of that is true, this style is bakcwards. I avoid it in Dream. See also the [Dream docs](https://aantron.github.io/dream/#type-error).

Compare with the definition of this type:

https://github.com/aantron/dream/blob/62d7cd48327bbc1d7ee5792c2d85547b995e31f9/src/pure/inmost.ml#L89-L118

in particular, note the final `;` on the last field

https://github.com/aantron/dream/blob/62d7cd48327bbc1d7ee5792c2d85547b995e31f9/src/pure/inmost.ml#L117-L118

which makes the whole `;`-first style pointless. The version in this PR is inconsistent with Dream, Web programming, and the OCamlFormat output on this type using this repo's config:

```ocaml
type error = {
  condition : [`Response of response | `String of string | `Exn of exn];
  layer : [`App | `HTTP | `HTTP2 | `TLS | `WebSocket];
  caused_by : [`Server | `Client];
  request : request option;
  response : response option;
  client : string option;
  severity : log_level;
  debug : bool;
  will_send_response : bool;
}
```

Reply:
I don't have strong opinion about style :+1:.